VideoLocker/src/main/java/org/edx/mobile/view/CourseDiscussionPostsThreadFragment.java

```diff
@@ -175,36 +176,41 @@ private boolean isFollowingTopics() {
         return DiscussionTopic.FOLLOWING_TOPICS.equalsIgnoreCase(discussionTopic.getName());
     }
 
-    private boolean listNeedsToBeRefreshed = true;
-
-    private void scheduleRefresh() {
-        if (isResumed()) {
-            populateThreadList(true);
-        } else {
-            listNeedsToBeRefreshed = true;
-        }
-    }
-
     public void onEventMainThread(DiscussionThreadFollowedEvent event) {
-        // TODO: Optimization: Only refresh the item if the followed post is currently listed
-        // TODO: Optimization: Only refresh the row that was updated, instead of the whole list
-        scheduleRefresh();
+        for (int i = 0; i < discussionPostsAdapter.getCount(); ++i) {
+            if (discussionPostsAdapter.getItem(i).getIdentifier().equalsIgnoreCase(event.getDiscussionThread().getIdentifier())) {
+                discussionPostsAdapter.replace(event.getDiscussionThread(), i);
+                break;
+            }
+        }
     }
 
     public void onEventMainThread(DiscussionCommentPostedEvent event) {
-        // TODO: Optimization: Only refresh if the comment is a reply to a currently listed post
-        // TODO: Optimization: Only refresh the row that was updated, instead of the whole list
-        scheduleRefresh();
+        for (int i = 0; i < discussionPostsAdapter.getCount(); ++i) {
+            final DiscussionThread discussionThread = discussionPostsAdapter.getItem(i);
+            if (discussionThread.getIdentifier().equalsIgnoreCase(event.getComment().getThreadId())) {
+                discussionThread.setCommentCount(discussionThread.getCommentCount() + 1);
+                discussionPostsAdapter.notifyDataSetChanged();
+                break;
+            }
+        }
     }
 
     public void onEventMainThread(DiscussionThreadPostedEvent event) {
-        // TODO: Optimization: Only refresh if the new thread belongs to this topic
-        scheduleRefresh();
+        // If a new post is created in this topic, insert it at the top of the list, after any pinned posts
+        if (!isFollowingTopics() && event.getDiscussionThread().getTopicId().equalsIgnoreCase(discussionTopic.getIdentifier())) {
+            int i = 0;
+            for (; i < discussionPostsAdapter.getCount(); ++i) {
+                if (!discussionPostsAdapter.getItem(i).isPinned()) {
+                    break;
+                }
+            }
+            discussionPostsAdapter.insert(event.getDiscussionThread(), i);
```

Comment:
Can we create a method for this inside the adapter itself? That seems to be neater than having this logic here.